src/bin/pg_autoctl/cli_show.c

```diff
@@ -400,12 +401,121 @@ cli_show_state(int argc, char **argv)
 }
 
 
+/*
+ * cli_show_local_state implements pg_autoctl show state --local, which
+ * composes the state from what we have in the configuration file and the state
+ * file for the local (keeper) node.
+ */
+static void
+cli_show_local_state()
+{
+	KeeperConfig config = keeperOptions;
+
+	switch (ProbeConfigurationFileRole(config.pathnames.config))
+	{
+		case PG_AUTOCTL_ROLE_MONITOR:
+		{
+			log_error("pg_autoctl show state --local is not supported "
+					  "on a monitor");
+			exit(EXIT_CODE_MONITOR);
+		}
+
+		case PG_AUTOCTL_ROLE_KEEPER:
+		{
+			bool missingPgdataIsOk = true;
+			bool pgIsNotRunningIsOk = true;
+			bool monitorDisabledIsOk = true;
+
+			Keeper keeper = { 0 };
+			CurrentNodeState nodeState = { 0 };
+
+			if (!keeper_config_read_file(&config,
+										 missingPgdataIsOk,
+										 pgIsNotRunningIsOk,
+										 monitorDisabledIsOk))
+			{
+				/* errors have already been logged */
+				exit(EXIT_CODE_BAD_CONFIG);
+			}
+
+			if (!keeper_init(&keeper, &config))
+			{
+				/* errors have already been logged */
+				exit(EXIT_CODE_BAD_CONFIG);
+			}
+
+			/* also grab the minimum recovery LSN if that's possible */
+			if (!pg_controldata(&(config.pgSetup), missingPgdataIsOk))
+			{
+				/* errors have already been logged, just continue */
+			}
+
+			/* build the CurrentNodeState from pieces */
+			nodeState.node.nodeId = keeper.state.current_node_id;
+			strlcpy(nodeState.node.name, config.name, _POSIX_HOST_NAME_MAX);
+			strlcpy(nodeState.node.host, config.hostname, _POSIX_HOST_NAME_MAX);
+			nodeState.node.port = config.pgSetup.pgport;
+
+			strlcpy(nodeState.formation, config.formation, NAMEDATALEN);
+			nodeState.groupId = config.groupId;
+
+			nodeState.reportedState = keeper.state.current_role;
+			nodeState.goalState = keeper.state.assigned_role;
+
+			strlcpy(nodeState.node.lsn,
+					config.pgSetup.control.recoveryEndingLocation,
+					PG_LSN_MAXLENGTH);
+
+			/* we have no idea, only the monitor knows, so display "?" */
+			nodeState.health = -1;
```

Comment:
I misunderstood the meaning of `Health` initially. We know the node is unhealthy if postgres is down then.

Reply:
Well, even when we're DEMOTED, we can't know what's the current health value on the monitor if we can't contact it. It might be stuck in the previous connection attempt at the moment when we do `pg_autoctl show state --local`, so I assume that we just don't know.